### PR TITLE
Fix extending parcelrc pipeline order

### LINKED ISSUE #4840: .parcelrc "transformer" key gives Uncaught SyntaxError: Cannot use import statement outside a module
# 🐛 bug report

## 🎛 Configuration (.babelrc, package.json, cli command)

```json
// .parcelrc
{
  "extends": ["@parcel/config-default"],
  "transformers": {}
}

// package.json

{
  "scripts": {
    "clean": "rm dist/bundle.js",
    "start": "parcel src/index.html",
    "build": "parcel build src/index.html",
    "test": "jest"
  },
  "dependencies": {
    "axios": "^0.19.2",
    "connected-react-router": "^6.8.0",
    "history": "^4.10.1",
    "immer": "^7.0.5",
    "normalizr": "^3.6.0",
    "querystring": "^0.2.0",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-redux": "^7.2.0",
    "react-router": "^5.2.0",
    "react-router-dom": "^5.2.0",
    "redux": "^4.0.5",
    "redux-devtools-extension": "^2.13.8",
    "redux-thunk": "^2.3.0",
    "reselect": "^4.0.0"
  },
  "alias": {
    "services": "./src/services",
    "state": "./src/state",
    "utils": "./src/utils",
    "components": "./src/components",
    "pages": "./src/pages",
    "assets": "./src/assets"
  },
  "devDependencies": {
    "@parcel/config-default": "^2.0.0-beta.1",
    "@parcel/transformer-babel": "^2.0.0-beta.1",
    "@parcel/transformer-js": "^2.0.0-beta.1",
    "@parcel/transformer-svg-react": "^2.0.0-beta.1",
    "@parcel/transformer-svgo": "^2.0.0-beta.1",
    "babel-jest": "^26.1.0",
    "eslint": "^7.3.1",
    "eslint-config-prettier": "^6.11.0",
    "eslint-plugin-prettier": "^3.1.4",
    "eslint-plugin-react": "^7.20.3",
    "express": "^4.17.1",
    "http-proxy-middleware": "^1.0.4",
    "jest": "^26.1.0",
    "parcel": "^2.0.0-beta.1",
    "prettier": "^2.0.5"
  }
}

// index.js

import React from "react";
import ReactDOM from "react-dom";

var mountNode = document.getElementById("app");
ReactDOM.render(<div>test</div>, mountNode);
```

## 🤔 Expected Behavior
It should not break the build

## 😯 Current Behavior
In the browser console I can see this error.
Throws this error: Uncaught SyntaxError: Cannot use import statement outside a module

## 🔦 Context
Was trying to add the "import svg as react component" capability to my project
I was getting the error described in in the "Current Behaviour" section.
Removed 
`"*.svg": ["@parcel/transformer-svg-react", "@parcel/transformer-babel", "@parcel/transformer-js"]` from the "transformers" object but I got the same problem. 

Initially, I thought that was a "merge" between the "extend" ("@parcel/config-default") and the mapping itself.
I've checked the "node_modules/@parcel/core/lib/requests/ParcelConfigRequest.js" file and it's `mergeConfigs` function seems to be working just fine.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | ^2.0.0-beta.1
| Node             | v13.13.0
| npm/Yarn         | 6.14.4
| Operating System | Pop!_OS 20.04 LTS

### PR DESCRIPTION
**Pull request by mischnic**
# ↪️ Pull Request

Closes https://github.com/parcel-bundler/parcel/issues/4856
Closes #4840

https://github.com/parcel-bundler/parcel/pull/4779 has broken parcelrc in a different way:
Even when specifying an empty transformers map in an extending config, all named pipelines get order to the front types`:*.{ts,tsx}, bundle-text:*, data-url:*, url:*, *.{js,mjs,jsm,jsx,es6,cjs,ts,tsx}, *.{json,json5} , ...`, causing a JS script tag with `url:index.`js to match the `url:*` pipeline….

So it looks like the reordering the transformers is the wrong approach and it should rather keep the order and prefer named pipelines when doing the lookup.

- Actually, this is still wrong (but because of the design of our config format):.
```jsonc
{
	// custom config
	"*.ts": ["@parcel/transformer-typescript-tsc"], 
	// extended config
	"types:*": ["@parcel/transformer-tsc-types"],
	"*.{js,ts}": ["@parcel/transformer-babel"],
	"url:*": ["@parcel/transformer-raw"]
}
```

1. ~When giving named pipelines precedence:~
`<script src="index.ts"></script>` requests `url:index.ts` and matches `url:*` (wrong)
`"types": "index.ts"` requests `types:index.ts` and matches `*.ts`  (correct)
`import "index.ts";` requests `index.ts` and matches `*.ts`  (correct)

2. ~Use first match (with or without pipeline)~
`<script src="index.ts"></script>` requests `url:index.ts` and matches `*.ts`  (correct)
`"types": "index.ts"` requests `types:index.ts` and match `*.ts` (wrong)
`import "index.ts";` requests `index.ts` and matches `*.ts`  (correct)

3. Use first match (pipeline needs to match exactly if specified) 👍 
~`<script src="index.ts"></script>` requests `url:index.ts` and matches `url:*` (wrong)~
`<script src="index.ts"></script>` requests `index.ts` and matches `*.ts` (correct)
`"types": "index.ts"` requests `types:index.ts` and match `types:*` (correct)
`import "index.ts";` requests `index.ts` and matches `*.ts`  (correct)

Somewhat related to that (at the moment neither approach 1,2 nor 3 is used)

```
true  === micromatch.isMatch('types:index.ts', '*.ts')		   ???
false === micromatch.isMatch('types:index.ts', '*.{ts}')
true  === micromatch.isMatch('types:index.ts', '*.{ts,tsx}')   ???
true  === micromatch.isMatch('index.ts',       '*.ts')
false === micromatch.isMatch('index.ts',       '*.{ts}')	   ???
true  === micromatch.isMatch('index.ts',       '*.{ts,tsx}')
```

- Should the `NAMED_PIPELINE_REGEX` test be done once and stored for every pipeline? Rather doing it again on every match?

## 💻 Examples

```json
{
	"extends": "@parcel/config-default",
	"transformers": {
		
	}
}
```

## 🚨 Test instructions

Parcelrc above with `<script src="index.js"></script>`

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.49s | +45.00ms |
| Cached | 2.27s | +7.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/index.html | 701.00b | +0.00b | 749.00ms | +100.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 744.00ms | +99.00ms ⚠️ |
| dist/legacy/index.0e158b6b.css | 77.00b | +0.00b | 1.43s | -246.00ms 🚀 |
| dist/modern/index.e480862d.css | 77.00b | +0.00b | 1.55s | +252.00ms ⚠️ |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 27.56s | -141.00ms |
| Cached | 2.42s | -61.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 147.00ms | -19.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 478.15kb | +0.00b | 220.00ms | -102.00ms 🚀 |
| dist/PermalinkedComment.c85d137d.js | 4.22kb | +0.00b | 215.00ms | -89.00ms 🚀 |
| dist/UserProfile.30e15baf.js | 1.62kb | +0.00b | 216.00ms | -88.00ms 🚀 |
| dist/NotFound.5aa6b041.js | 387.00b | +0.00b | 216.00ms | -87.00ms 🚀 |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 153.00ms | -86.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.15m | -6.41s |
| Cached | 5.62s | +379.00ms ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.8fef7fe2.js | 2.73mb | -3.00b 🚀 | 1.82m | -3.75s |
| dist/popup.4a66e15e.js | 170.62kb | +0.00b | 58.66s | +21.50s ⚠️ |
| dist/EmojiPickerComponent.bab709db.js | 97.89kb | +0.00b | 14.12s | +989.00ms ⚠️ |
| dist/card.ce67dc35.js | 55.09kb | +0.00b | 15.07s | -915.00ms 🚀 |
| dist/Modal.e3bd77d0.js | 41.40kb | +0.00b | 5.86s | -959.00ms 🚀 |
| dist/component.59edff1f.js | 30.78kb | +0.00b | 7.50s | +4.03s ⚠️ |
| dist/component.37867f50.js | 22.51kb | +0.00b | 5.85s | -975.00ms 🚀 |
| dist/DatePicker.cbe44a9b.js | 21.17kb | +0.00b | 13.10s | +1.81s ⚠️ |
| dist/js.48be70a4.js | 16.40kb | +0.00b | 5.85s | -975.00ms 🚀 |
| dist/workerHasher.65b8a1b7.js | 11.78kb | +0.00b | 1.05m | +48.06s ⚠️ |
| dist/EmojiTypeAheadComponent.9af6e5fc.js | 8.67kb | +0.00b | 1.05m | +49.30s ⚠️ |
| dist/component.11476a92.js | 6.05kb | +0.00b | 5.85s | -975.00ms 🚀 |
| dist/png-chunks-extract.8d15a914.js | 3.41kb | +0.00b | 5.85s | -975.00ms 🚀 |
| dist/Modal.2835744c.js | 3.08kb | +0.00b | 7.44s | +2.09s ⚠️ |
| dist/16.d1b8e6f7.js | 2.38kb | +0.00b | 1.18s | -1.50s 🚀 |
| dist/EmojiUploadComponent.88a444df.js | 2.18kb | +0.00b | 1.05m | +49.30s ⚠️ |
| dist/ResourcedEmojiComponent.d9a49606.js | 2.00kb | +0.00b | 7.44s | +2.10s ⚠️ |
| dist/date.d49e64f8.js | 1.73kb | +0.00b | 6.75s | -659.00ms 🚀 |
| dist/16.f40c1329.js | 1.73kb | +0.00b | 6.32s | -1.03m 🚀 |
| dist/images.80270b1f.js | 1.67kb | +0.00b | 6.89s | -512.00ms 🚀 |
| dist/16.86a8c7d4.js | 1.66kb | +0.00b | 6.09s | -840.00ms 🚀 |
| dist/feedback.8b7c7928.js | 1.64kb | +0.00b | 9.08s | +1.23s ⚠️ |
| dist/workerHasher.9bb425b1.js | 1.54kb | +0.00b | 1.05m | +49.30s ⚠️ |
| dist/list-number.4632b459.js | 1.46kb | +0.00b | 7.11s | -418.00ms 🚀 |
| dist/status.cccf04be.js | 1.46kb | +0.00b | 7.27s | -447.00ms 🚀 |
| dist/16.cff410ff.js | 1.38kb | +0.00b | 6.30s | -766.00ms 🚀 |
| dist/code.af9ca2b3.js | 1.38kb | +0.00b | 6.75s | -519.00ms 🚀 |
| dist/16.c1ed07f8.js | 1.33kb | +0.00b | 6.09s | -840.00ms 🚀 |
| dist/16.6603f2f3.js | 1.33kb | +0.00b | 6.09s | -980.00ms 🚀 |
| dist/16.a4b392c2.js | 1.32kb | +0.00b | 1.18s | -1.48s 🚀 |
| dist/16.249c644a.js | 1.31kb | +0.00b | 6.54s | -728.00ms 🚀 |
| dist/link.d12cc130.js | 1.30kb | +0.00b | 6.89s | -630.00ms 🚀 |
| dist/heading6.4c4017db.js | 1.30kb | +0.00b | 9.08s | +1.23s ⚠️ |
| dist/heading3.93606b9c.js | 1.29kb | +0.00b | 7.46s | -387.00ms 🚀 |
| dist/16.ab5c0088.js | 1.27kb | +0.00b | 6.54s | -732.00ms 🚀 |
| dist/emoji.e603e70e.js | 1.23kb | +0.00b | 6.89s | -512.00ms 🚀 |
| dist/16.afee1494.js | 1.23kb | +0.00b | 6.54s | -1.02m 🚀 |
| dist/16.e60b4d77.js | 1.21kb | +0.00b | 1.18s | -1.48s 🚀 |
| dist/16.7c7a40c9.js | 1.19kb | +0.00b | 6.54s | -1.02m 🚀 |
| dist/16.06d5a67b.js | 1.18kb | +0.00b | 6.75s | -520.00ms 🚀 |
| dist/heading5.dde15cec.js | 1.17kb | +0.00b | 9.08s | +1.23s ⚠️ |
| dist/16.2096b9f3.js | 1.14kb | +0.00b | 1.18s | -1.48s 🚀 |
| dist/16.f040d62d.js | 1.14kb | +0.00b | 6.30s | -767.00ms 🚀 |
| dist/16.ecf8a702.js | 1.13kb | +0.00b | 6.09s | -860.00ms 🚀 |
| dist/16.1b342686.js | 1.13kb | +0.00b | 6.30s | -768.00ms 🚀 |
| dist/16.2e7f52c6.js | 1.13kb | +0.00b | 6.30s | -767.00ms 🚀 |
| dist/16.da43aa07.js | 1.13kb | +0.00b | 6.09s | -840.00ms 🚀 |
| dist/16.dd07505e.js | 1.10kb | +0.00b | 1.18s | -4.16s 🚀 |
| dist/component.40194190.js | 1.07kb | +0.00b | 7.44s | +2.10s ⚠️ |
| dist/mention.b1828447.js | 1.07kb | +0.00b | 7.11s | -418.00ms 🚀 |
| dist/heading4.5316f6f2.js | 1.06kb | +0.00b | 7.46s | -386.00ms 🚀 |
| dist/16.317b9164.js | 1.05kb | +0.00b | 6.54s | -1.02m 🚀 |
| dist/layout.ac6734f9.js | 1.05kb | +0.00b | 6.89s | -630.00ms 🚀 |
| dist/divider.d2be55a1.js | 1.03kb | +0.00b | 6.89s | -512.00ms 🚀 |
| dist/quote.f4e09143.js | 1.02kb | +0.00b | 7.27s | -447.00ms 🚀 |
| dist/action.cb287eb0.js | 1.00kb | +0.00b | 6.75s | -520.00ms 🚀 |
| dist/decision.91f3af14.js | 1010.00b | +0.00b | 6.75s | -651.00ms 🚀 |
| dist/list.057cb664.js | 978.00b | +0.00b | 7.11s | -418.00ms 🚀 |
| dist/panel-error.c4b05f6f.js | 906.00b | +0.00b | 7.11s | -520.00ms 🚀 |
| dist/panel-note.ec8386b7.js | 841.00b | +0.00b | 7.11s | -520.00ms 🚀 |
| dist/EmojiPickerComponent.a968dafe.js | 535.00b | +0.00b | 9.10s | -4.43s 🚀 |
| dist/index.html | 119.00b | +0.00b | 1.04s | +82.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.8fef7fe2.js | 2.73mb | +4.00b ⚠️ | 671.00ms | +19.00ms |

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/270525386471768589)**

**Comment by mischnic** [comment]
The new logic is:

1. Don't reorder pipelines when merging configs
2. When matching a pipeline to a request:
	 -  find a pipeline that matches the request's pipeline prefix (either same prefix or both have no named pipeline) and matches the glob, throw otherwise
	- get a list of all pipelines that match without the request's pipeline to resolve any eventual `, "..."`

File: packages/core/core/src/ParcelConfig.js [hunk 1]
```diff
@@ -333,10 +333,16 @@ export default class ParcelConfig {
   }
 
   isGlobMatch(filePath: FilePath, pattern: Glob, pipeline?: ?string) {
-    let prefix = pipeline ? `${pipeline}:` : '';
+    let [patternPipeline, patternGlob] = pattern.split(':');
+    if (!patternGlob) {
+      patternGlob = patternPipeline;
+      patternPipeline = null;
+    }
```
Comment (mischnic) [inline packages/core/core/src/ParcelConfig.js:340]: I guess this be stored in the config and not recalculated every time?

**Comment by 101arrowz** [comment]
I don't know if this is related, but it seems to be impossible to have configs actually merge. I have two configs, one that adds a `'service-worker'` runtime and another that adds a few transformers and a namer. Whenever I extend both configs (and `@parcel/config-default`) one of them is overridden and the changes are not applied. It seems that this is the result of the defaults in the `ParcelConfig` constructor, but I'm not sure.

In addition, the globs should be expanded before merging configs, i.e. Parcel should transform
```json
{
    "transformers": {
        "*.{htm,html}": ["@parcel/transformer-html"]
    }
}
```
to 
```json
{
    "transformers": {
        "*.htm": ["@parcel/transformer-html"],
        "*.html": ["@parcel/transformer-html"]
    }
}
```
That's important because a config may override the transformer for just `*.htm` and not `*.html`, but right now because of the globs the override won't work.

**Comment by mischnic** [comment]
Please open a new issue with all your config files + parcelrc

**Comment by 101arrowz** [comment]
This could be better suited in a separate PR, but the globs should be expanded before merging configs, i.e. Parcel should convert
```json
{
    "transformers": {
        "*.{htm,html}": ["@parcel/transformer-html"]
    }
}
```
to 
```json
{
    "transformers": {
        "*.htm": ["@parcel/transformer-html"],
        "*.html": ["@parcel/transformer-html"]
    }
}
```
That's important because a config may override the transformer for just `*.htm` and not `*.html`, but right now because of the globs the override won't work.

**Comment by mischnic** [comment]
> This could be better suited in a separate PR

1. Commenting about the same issue in multiple issues/PRs makes it hard to follow the discussion
2. Commenting on a PR about a issue is generally a bad idea, since the discussion becomes basically invisible once the PR gets merged

File: packages/core/core/src/ParcelConfig.js (hunk 1 above)
Comment (devongovett) [reply packages/core/core/src/ParcelConfig.js:380]: Yeah we can also compile the globs to regexes in advance

File: packages/transformers/typescript-types/src/TSTypesTransformer.js [hunk 2]
```diff
@@ -155,6 +155,8 @@ export default (new Transformer({
     return [
       {
         type: 'ts',
+        // Stay on the types pipeline, even if the type changes
+        pipeline: asset.pipeline,
```
Comment (mischnic) [inline packages/transformers/typescript-types/src/TSTypesTransformer.js:159]: Due to a bug in Transformation, this never surfaced, but changing an asset type from `tsx` to `ts` would also reset the pipeline, and this Typescript asset would then continue on the regular js pipeline.

**Comment by mischnic** [comment]
Transformation didn’t take into account that asset.value.pipeline can change over the course of the transformation, which is why Vue was broken. A consequence of that fix is also https://github.com/parcel-bundler/parcel/pull/4842#discussion_r511660315

**Comment by 101arrowz** [comment]
> `Transformation.js` didn’t take into account that asset.value.pipeline can change over the course of the transformation, **which is why Vue was broken**. A consequence of that fix is also [#4842 (comment)](https://github.com/parcel-bundler/parcel/pull/4842#discussion_r511660315)

I don't believe the Vue transformer changes the pipeline of an asset at any point, but rather generates code that imports other assets with different pipelines. On a side note, we might also consider redoing the transformer with query parameters, since [that's what Evan You himself suggests](https://github.com/parcel-bundler/parcel/issues/3364#issuecomment-622086191).

**Comment by mischnic** [comment]
The `pipeline`/`type` changes when going from the `foo.vue` asset to the `style:foo.vue` child asset.
